Index: src/interface.c
===================================================================
--- src/interface.c	(revision 171)
+++ src/interface.c	(working copy)
@@ -1395,7 +1395,7 @@
         {
             home_dir=getenv("HOME");
             strcpy(hfile_path,home_dir);
-            strcat(hfile_path,"/.m17n/");
+            strcat(hfile_path,"/.m17n.d/");
             strncat(hfile_path,mfile,strlen(mfile));
             strcat(hfile_path,".mim");
 
@@ -1791,7 +1791,7 @@
     cnt=0;
     home_dir=getenv("HOME");
     strcpy(custom_dir_path,home_dir);
-    strcat(custom_dir_path,"/.m17n/");
+    strcat(custom_dir_path,"/.m17n.d/");
 
     n = scandir(custom_dir_path, &namelist, 0, alphasort);
     if (n < 0)
@@ -2736,7 +2736,7 @@
     lc=get_iso_code(current_mim);
     sprintf(imname,"(input-method %s %s-inscript)\n\n",lc,filenameprefix);
     strcpy(hdirpath,home_dir);
-    strcat(hdirpath,"/.m17n/");
+    strcat(hdirpath,"/.m17n.d/");
 
     if ( stat (hdirpath , &buf ) != 0 )
     {
@@ -2747,7 +2747,7 @@
 
     }
 
-    sprintf(nfilepath,"%s/.m17n/%s-%s-inscript.mim",home_dir,lc,filenameprefix);
+    sprintf(nfilepath,"%s/.m17n.d/%s-%s-inscript.mim",home_dir,lc,filenameprefix);
     nfilepath[strlen(nfilepath)]='\0';
     fpm = fopen(nfilepath, "w+");
 
@@ -2904,7 +2904,7 @@
 
     home_dir=getenv("HOME");
     strcpy(sdirpath,home_dir);
-    strcat(sdirpath,"/.m17n/");
+    strcat(sdirpath,"/.m17n.d/");
 
     if ( stat (sdirpath , &buf ) != 0 )
     {
